lego/apps/users/models.py

```diff
@@ -495,26 +519,69 @@ class Penalty(BasisModel):
     source_event = models.ForeignKey(
         "events.Event", related_name="penalties", on_delete=models.CASCADE
     )
+    activation_time = models.DateTimeField("date created", null=True, default=None)
 
     objects = UserPenaltyManager()  # type: ignore
 
-    def expires(self):
-        dt = Penalty.penalty_offset(self.created_at) - (
-            timezone.now() - self.created_at
+    def save(self, *args, **kwargs):
+        last_penalty_to_be_expired = Penalty.objects.filter(
+            user=self.user, activation_time__lte=timezone.now()
+        ).order_by("-activation_time")
+
+        self.activation_time = (
+            last_penalty_to_be_expired[0].exact_expiration
+            if len(last_penalty_to_be_expired) != 0
+            else self.created_at
         )
-        return dt.days
+        super().save(*args, **kwargs)
+
+    def default_save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+
+    def expires(self):
+        expirationDate = Penalty.penalty_offset(
+            self.activation_time, weight=self.weight
+        ) - (timezone.now() - self.activation_time)
+        return expirationDate.days
+
+    def delete_and_adjust_future_activation_times(self):
+        if self.weight == 1:
+            new_activation_time = timezone.now()
+            future_penalties = Penalty.objects.filter(
+                user=self.user, activation_time__gt=new_activation_time
+            ).order_by("activation_time")
+
+            for penalty in future_penalties:
+                penalty.activation_time = new_activation_time
+                penalty.default_save()
+                new_activation_time = penalty.exact_expiration
+
+            self.delete()
+        else:
+            self.weight -= 1
+            self.activation_time = timezone.now()
+            future_penalties = Penalty.objects.filter(
+                user=self.user, activation_time__gt=self.activation_time
+            ).order_by("activation_time")
+            new_activation_time = self.exact_expiration
+            self.default_save()
+
+            for penalty in future_penalties:
+                penalty.activation_time = new_activation_time
+                penalty.default_save()
+                new_activation_time = penalty.exact_expiration
 
     @property
     def exact_expiration(self):
```

Comment:
i know that you did not choose this name, but it kind of annoys me that this is called exect_expiration and not just expiration... But it would have to be changed in the front end too, and it really doesn't matter in any other way than it anoying me, so it's fine

Reply:
I agree and i can fix this but in another pull request :)